include/gatekeeper_fib.h

```diff
@@ -102,8 +102,13 @@ struct ether_cache {
 	/* The IP address of the nexthop. */
 	struct ipaddr    ip_addr;
 
-	/* The whole Ethernet header. */
-	struct ether_hdr eth_hdr;
+	/* The whole link-layer header. */
+	struct {
+		/* Ethernet header (required). */
+		struct ether_hdr eth_hdr;
+		/* VLAN header (optional). */
+		struct vlan_hdr  vlan_hdr;
+	} l2_hdr;
```

Comment:
Given that we use `l2_hdr` as a packet header, that is, we expect no gap between `eth_hdr` and `vlan_hdr`, we should add `__attribute__((__packed__));` between `}` and `l2_hdr;`